### PR TITLE
[modbus.sbc] Initial contribution

### PR DESCRIPTION
**Pull request by fwolter**
Signed-off-by: Fabian Wolter <github@fabian-wolter.de>

### CODE REVIEW

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java:87]: Unnecessary? You could construct QuantityType using the DecimalType without intermediate conversion to float

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java:33]: Same index as above?

File: bundles/org.openhab.binding.modbus.sbc/README.md
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/README.md:22]: This is already specified in the parent serial thing, I think it makes sense to use it from there and remove this parameter.

File: bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/config/config.xml
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/config/config.xml:12]: Remove (see comment above)

File: bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/config/config.xml
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/config/config.xml:16]: Add unit information

**Review by ssalonen — CHANGES_REQUESTED** [review]
Good job, very clean contribution and could act as a template for the simple read-only modbus bindings.

Some changes suggested.

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java:70]: I'm a bit confused, too. The documentation states "In accordance with the Modbus protocol, a register R is numbered as R - 1 when transmitted." page 5 https://www.meterbuy.com/fileadmin/user_upload/Data_sheets/120102_SAIA_-_Data_sheet_26-529_EN_DS_Energy-Meter-ALD1-with-Modbus.pdf

Is this common to Modbus devices?

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java:87]: I need to apply a multiplier before converting it to QuantityType. Do you have a suggestion?

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java]: They are for distinct Thing types. I added a comment.

File: bundles/org.openhab.binding.modbus.sbc/README.md
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/README.md]: Great hint! To retrieve the slave address I modified the BaseModbusThingHandler. This was missing in the current implementation.

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java:70]: In modbus binding terminology the `register address` refers to the address number on the wire, i.e. `R-1`.

This is typical source of confusion with Modbus. I even added section in the documentation clarifying the topic: https://www.openhab.org/addons/bindings/modbus/#comment-on-addressing

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java:88]: Would this work?

```suggestion
                        .map(decimalState -> new QuantityType(decimalState.toBigDecimal().multiply(new BigDecimal(multiplier)), channel.getUnit()))
```

It's a minor cosmetic change to the original but avoids floating point conversions.

**Review by ssalonen — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java]: I wasn't aware of this fine difference in wording. Changed it to `FIRST_READ_REGISTER`.

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java]: Yes, I'm staying in the BigDecimal domain, now.

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java]: I would have suggested to keep everything zero based, similar to the other bindings 😅...adding simply a comment referencing device manual

I think this is OK as well but perhaps good to add comment for posterity that this is 1-based register number, not the zero based address that goes on the wire.

Also note the same naming issue with channel definitions.

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java:48]:
```suggestion
    TOTAL_ENERGY("0.01", 28, UINT32, SmartHomeUnits.KILOWATT_HOUR),
    PARTIAL_ENERGY("0.01", 30, UINT32, SmartHomeUnits.KILOWATT_HOUR), // only unidirectional meters
    FEEDING_BACK_ENERGY("0.01", 30, UINT32, SmartHomeUnits.KILOWATT_HOUR), // only bidirectional meters
    VOLTAGE("1", 36, UINT16, SmartHomeUnits.VOLT),
    CURRENT("0.1", 37, UINT16, SmartHomeUnits.AMPERE),
    ACTIVE_POWER("10", 38, INT16, SmartHomeUnits.WATT),
    REACTIVE_POWER("10", 39, INT16, SmartHomeUnits.VAR),
    POWER_FACTOR("0.01", 40, UINT16, SmartHomeUnits.ONE);

    private BigDecimal multiplier;
    private int registerAddress;
    private ModbusConstants.ValueType type;
    private Unit<?> unit;

    private ALD1Registers(String multiplier, int registerAddress, ValueType type, Unit<?> unit) {
        this.multiplier = new BigDecimal(multiplier);
```

Just a cosmetic suggestion 

With this change you would have "neat" finite numbers since big decimal operate in 10-base. See https://stackoverflow.com/questions/7408566/java-double-value-0-01-changes-to-0-009999999999999787 

Feel free to ignore this comment :)

Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java:48]: I understand, but it seems more odd to me to have these numeric constants as strings...

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Handler.java]: Actually I always use zero-based numbers where possible, but my goal here was to use the same numbers in the constants as in the device's documentation. I added a comment and changed the wording in the enum.

File: bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.sbc/src/main/java/org/openhab/binding/modbus/sbc/internal/ALD1Registers.java]: Yup.. No way around the strings really if we want stay fully in BigDecimal domain. 

As said, feel free to ignore this comment, makes little practical difference really

**Comment by ssalonen** [comment]
Lgtm!

**Review by ssalonen — APPROVED** [review]

**Review by cpmeister — APPROVED** [review]
LGTM

**Comment by fwolter** [comment]
@openhab/add-ons-maintainers As @cpmeister reviewed this PR and I created it, we need another reviewer here.

File: bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/thing/thing-types.xml
Comment (andrewfg) [inline bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/thing/thing-types.xml:10]: You may want to consider adding support for the RTU over TCP ( #9435 ) by adding the following here..

```
<bridge-type-ref id="tcp"/>
```

Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/thing/thing-types.xml:10]: I agree. People can use serial/ethernet gateways (passing modbus rtu over tcp) or modbus gateways (converting modbus rtu to modbus tcp) with these devices

Comment (fwolter) [reply bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/thing/thing-types.xml:10]: I wasn't aware this is possible. Thanks!

**Comment by andrewfg** [comment]
FYI: The “RTU over TCP” revisions have now been merged. So this one should be able to build again.

File: bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Hilbrand) [reply bundles/org.openhab.binding.modbus.sbc/src/main/resources/OH-INF/thing/thing-types.xml:10]: Is adding `tcp` bridge something that will be added here?

**Review by Hilbrand — COMMENTED** [review]
Code looks ok. Just needs update of copyright year and maybe adding that tcp bridge?

**Comment by fwolter** [comment]
Forgot to add it to the unidirectional meter, too. Thanks! The license headers are also fixed.

**Review by Hilbrand — APPROVED** [review]
LGTM

**Pull request merged**
